### PR TITLE
Fix GCStress assertion

### LINKED ISSUE #17598: [arm32/Ubuntu] GCStress failures: "Consistency check failed: FAILED: !pThread->HasPendingGCStressInstructionUpdate()"
When I run GCStress=0xc on Ubuntu arm32, the most common failure is:

```
               Assert failure(PID 23543 [0x00005bf7], Thread: 23543 [0x5bf7]): Consistency check failed: FAILED: !pThread->HasPendingGCStressInstructionUpdate()
                   File: /mnt/coreclr/src/vm/gccover.cpp Line: 1772
                   Image: /ssd/brucefo/checked/coreclr/bin/tests/Linux.arm.Checked/Tests/Core_Root/corerun

               ./30032.sh: line 243: 23543 Aborted                 $_DebuggerFullPath "$CORE_ROOT/corerun" $ExePath $CLRTestExecutionArguments
               Expected: 100
               Actual: 134
               END EXECUTION - FAILED
```

The most common failures are the tests in these directories:
```
baseservices/threading/generics/TimerCallback
baseservices/threading/generics/WaitCallback
```

as well as
```
baseservices/threading/regressions/beta2/437044/437044.sh
```
which in at least one of my runs required 5 retries to pass.

This is https://github.com/dotnet/coreclr/issues/5310.

### PR DESCRIPTION
**Pull request by sdmaclea**
Fixes #17598

@AndyAyersMS @BruceForstall PTAL

### CODE REVIEW

File: src/vm/gccover.cpp [hunk 1]
```diff
@@ -1728,7 +1728,12 @@ void DoGcStress (PCONTEXT regs, MethodDesc *pMD)
     // BUG(github #10318) - when not using allocation contexts, the alloc lock
     // must be acquired here. Until fixed, this assert prevents random heap corruption.
     assert(GCHeapUtilities::UseThreadAllocationContexts());
-    if (!GCHeapUtilities::GetGCHeap()->StressHeap(GetThread()->GetAllocContext()))
+    GCHeapUtilities::GetGCHeap()->StressHeap(GetThread()->GetAllocContext());
+
+    // StressHeap can exit early w/o forcing a SuspendEE to trigger the instruction update
+    // We can not rely on the return code to determine if the instruction update happened
```
Comment (sdmaclea) [inline src/vm/gccover.cpp]: StressHeap() will call into GarbageCollectGeneration() which has an early out 
```
    //don't trigger another GC if one was already in progress
    //while waiting for the lock
```

File: src/vm/threads.h [hunk 2]
```diff
@@ -4838,8 +4838,8 @@ class Thread: public IUnknown
         LIMITED_METHOD_CONTRACT;
         PRECONDITION(HasPendingGCStressInstructionUpdate());
 
-        m_pbDestCode = NULL;
-        m_pbSrcCode = NULL;
+        VolatileStoreWithoutBarrier<BYTE*>(&m_pbDestCode, NULL);
+        VolatileStore<BYTE*>(&m_pbSrcCode, NULL);
```
Comment (sdmaclea) [inline src/vm/threads.h]: This guarantees the clear of dest occurs (is observed) before src, for similar reasons.

**Comment by AndyAyersMS** [comment]
I think this looks reasonable.

If the EE is not really suspended then we're making an on-the-fly instruction stream update which is a bit risky, but we do that elsewhere too.

**Comment by sdmaclea** [comment]
>  we're making an on-the-fly instruction stream update
Note `UpdateGCStressInstructionWithoutGC()` function is implemented as 
```
SuspendEE()
ResumeEE()
```

**Comment by sdmaclea** [comment]
@BruceForstall ping

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**